### PR TITLE
oobmigration: Preload additional result chunks in SCIP migrator

### PR DESCRIPTION
**Pull request by efritz**
Improvements to the SCIP migrator prior to its registration in https://github.com/sourcegraph/sourcegraph/pull/45106. This PR attempts to preload/cache-warm definitions of the document to be processed _next_. This made the migration of the dotcom database quite a bit faster as we had fewer round-trip queries of result chunks for smaller repositories.

## Test plan

Tested by hand.

### CODE REVIEW

File: enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go [hunk 1]
```diff
@@ -211,18 +211,31 @@ func migrateUpload(
...
 				resultChunkCache,
+				cacheSize,
 				uploadID,
 				numResultChunks,
 				indexerName,
 				path,
-				document,
+				documentsByPath[path],
+				definitionResultIDs[i],
+				definitionResultIDs[i+1:],
```
Comment (efritz) [inline enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:238]:
```suggestion
				// Load all of the definitions for this document
				definitionResultIDs[i],
				// Load as many definitions from the next document as possible
				definitionResultIDs[i+1:],
```

File: enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go [hunk 2]
```diff
@@ -343,30 +361,41 @@ func fetchResultChunks(
...
 	indexMap := map[int]struct{}{}
 
-	for _, id := range ids {
-		// Calculate result chunk index that this identifier belongs to
-		idx := precise.HashKey(precise.ID(id), numResultChunks)
+outer:
+	for i, ids := range append([][]ID{ids}, preloadDefinitionResultIDs...) {
+		for _, id := range ids {
+			if len(indexMap) >= resultChunkCacheSize && i != 0 {
+				break outer
```
Comment (efritz) [inline enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:377]:
```suggestion
				// Only add fetch preload IDs if we have more room in our request
				break outer
```

File: enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go [hunk 3]
```diff
@@ -343,30 +361,41 @@ func fetchResultChunks(
...
+			// Attempt to load result chunk data from the cache. If it's present then we can add it to
+			// the output map immediately. If it's not present in the cache, then we'll need to fetch it
+			// from the database. Collect each such result chunk index so we can do a batch load.
 
-		if rawResultChunk, ok := resultChunkCache.Get(idx); ok {
-			resultChunks[idx] = rawResultChunk.(ResultChunkData)
-		} else {
-			indexMap[idx] = struct{}{}
+			if rawResultChunk, ok := resultChunkCache.Get(idx); ok {
+				if i == 0 {
```
Comment (efritz) [inline enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:393]:
```suggestion
				if i == 0 {
					// Don't stash preloaded result chunks for return
```

File: enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go [hunk 4]
```diff
@@ -378,30 +398,53 @@ func fetchResultChunks(
...
 	serializer *serializer,
 	resultChunkCache *lru.Cache,
+	resultChunkCacheSize int,
 	uploadID int,
 	numResultChunks int,
 	ids []ID,
+	preloadDefinitionResultIDs [][]ID,
 ) (map[int]ResultChunkData, error) {
+	// Stores a set of indexes that need to be loaded from the database. The value associated
+	// with an index is true if the result chunk should be returned to teh caller and false if
```
Comment (Numbers88s) [inline enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:408]: lookie lookie, I see a typo.

```suggestion
	// with an index is true if the result chunk should be returned to the caller and false if
```

Comment (efritz) [reply enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:408]: teh :clueless:

**Review by Numbers88s — APPROVED** [review]

**Pull request merged**